Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp

```diff
@@ -16,23 +31,61 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return totalAmpHoursUsed_;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    return (currentNow_ < 0);
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+   double hoursTillChargedOrDepleted, msTillChargedorDepleted;
+   int hours, minutes, seconds, milliseconds;
+
+   if (isCharging())
+        hoursTillChargedOrDepleted = abs(totalAmpHoursUsed_ / currentNow_);
+   else
+        hoursTillChargedOrDepleted = (BATTERY_AMP_HOUR_CAPACITY - totalAmpHoursUsed_) / currentNow_;
+
+   while (hoursTillChargedOrDepleted > 24)
+        hoursTillChargedOrDepleted -= 24;
+
+   msTillChargedorDepleted = hoursTillChargedOrDepleted * MS_IN_HOURS;
+
+   hours = floor(hoursTillChargedOrDepleted);
+   msTillChargedorDepleted -= hours * MS_IN_HOURS;
+
+   minutes = floor(msTillChargedorDepleted / MS_IN_MINUTES);
+   msTillChargedorDepleted -= minutes * MS_IN_MINUTES;
+
+   seconds = floor(msTillChargedorDepleted / MS_IN_SECONDS);
+   msTillChargedorDepleted -= seconds * MS_IN_SECONDS;
+
+   milliseconds = floor(msTillChargedorDepleted);
+
+   QTime timeRemaining(hours, minutes, seconds, milliseconds);
+   return timeRemaining;
```

Comment:
The logic here should be placed inside addData, and this can just be a getter of a private variable.